### PR TITLE
[shopfloor] rework datamodel

### PR DESCRIPTION
**Pull request by guewen**
This is a simplification of the model, which was hard to understand
and configure.
We had too many options to filter the visibility of menus / processes,
which made things hard to debug. Some models were overlapping in term
of concept and fonctionality.

* Processes and menus are redundant, remove processes and keep only menus
* Remove shopfloor groups, we can restrict the visibility of a menu
  using profiles. Later, only a manager should be able to change the
  profile on the devices.
  Additionally, we would still be able to use ir.rules if needed.
* On the menus, we'll be able to select several picking types. When a scenario has to know a single
  picking type, for instance to create a new move, we'll use a dedicated
  many2one for this purpose.

Details in commits

### CODE REVIEW

**Comment by guewen** [comment]
@simahawk In theory, it shouldn't have any impact on the front-end side (except the changes I already did). I tested the config / profiles / menus of the app, which still work, but didn't run any scenario yet.

**Comment by simahawk** [comment]
I'm about to review. Can you review this 1st https://github.com/camptocamp/wms/pull/9 ?
You can pay attention to backend part only as I fixed few issues coming from recent changes (tests were not updated as I did it in a rush).
Would be nice to merge and rebase here then.

File: shopfloor/models/stock_quant_package.py [hunk 1]
```diff
@@ -4,6 +4,13 @@
 class StockQuantPackage(models.Model):
     _inherit = "stock.quant.package"
 
+    move_line_ids = fields.One2many(
```
Comment (simahawk) [inline shopfloor/models/stock_quant_package.py:7]: is this needed if we rely on https://github.com/camptocamp/wms/pull/9/commits/0ac9e9f4db76b3df00ef42d8563f4430984abfb1#diff-8669ee167ce03adadc1c12257e201f58R41 ?

File: shopfloor/__manifest__.py [hunk 2]
```diff
@@ -25,9 +25,7 @@
     ],
     "data": [
         "security/ir.model.access.csv",
-        "views/shopfloor_operation_group.xml",
```
Comment (simahawk) [inline shopfloor/__manifest__.py:28]: great! Is the update of our integration module on your radar?

File: shopfloor/services/checkout.py [hunk 3]
```diff
@@ -85,17 +85,19 @@ def scan_document(self, barcode):
...
             if package:
-                pickings = (
-                    self.env["stock.move.line"]
-                    .search(
-                        [
-                            ("state", "not in", ("cancel", "done")),
-                            ("package_id", "=", package.id),
-                            ("picking_id.picking_type_id", "=", self.picking_type.id),
-                        ]
+                pickings = package.move_line_ids.filtered(
```
Comment (simahawk) [inline shopfloor/services/checkout.py:88]: TBH I prefer the search, I factored it out to a specific helper method in my PR

File: shopfloor/services/menu.py [hunk 4]
```diff
@@ -21,21 +20,18 @@ class ShopfloorMenu(Component):
...
-                ("operation_group_ids", "in", profile.operation_group_ids.ids),
+        # FIXME exclude menus if they have no picking types
+        # in warehouse matching the profile
+        return expression.AND(
+            [
+                base_domain,
+                [
+                    "|",
+                    ("profile_ids", "=", False),
+                    ("profile_ids", "in", self.work.profile.id),
```
Comment (simahawk) [inline shopfloor/services/menu.py]:
```suggestion
                    ("profile_ids", "in", self.work.profile.ids),
```

File: shopfloor/services/service.py [hunk 5]
```diff
@@ -24,19 +24,21 @@ class BaseShopfloorService(AbstractComponent):
...
-        """
-        Get the current picking type based on the menu and the warehouse of the profile.
-        """
-        picking_type = self.work.menu.process_id.picking_type_id
-        if picking_type.warehouse_id != self.work.profile.warehouse_id:
+    def picking_types(self):
+        """Return picking types for the menu and profile"""
+        # TODO make this a lazy property or computed field avoid running the
+        # filter every time?
+        picking_types = self.work.menu.picking_type_ids.filtered(
```
Comment (simahawk) [inline shopfloor/services/service.py:31]: shall we cache this by menu/profile?

File: shopfloor/services/single_pack_putaway.py [hunk 6]
```diff
@@ -85,8 +85,12 @@ def _response_for_start_success(self, move_line, pack):
 
     def start(self, barcode):
         """Scan a pack barcode"""
-
-        picking_type = self.picking_type
+        # TODO we have to rework this and single_pack_transfer, 'pack putaway'
```
Comment (simahawk) [inline shopfloor/services/single_pack_putaway.py:88]: :dizzy_face:

File: shopfloor/tests/test_checkout_cancel_line.py [hunk 7]
```diff
@@ -85,7 +85,10 @@ def test_cancel_package_ok(self):
         self.assert_response(
             response,
             next_state="summary",
-            data={"picking": self._stock_picking_data(picking, done=True)},
+            data={
+                "picking": self._stock_picking_data(picking, done=True),
+                "all_processed": False,
```
Comment (simahawk) [inline shopfloor/tests/test_checkout_cancel_line.py:90]: doh, sorry, I fixed these already :sweat_smile:

File: shopfloor/models/stock_quant_package.py (hunk 1 above)
Comment (guewen) [reply shopfloor/models/stock_quant_package.py:7]: I try to limit the number of calls to search which never use the cache

File: shopfloor/services/menu.py [hunk 8]
```diff
@@ -21,31 +20,35 @@ class ShopfloorMenu(Component):
...
-            ("operation_group_ids.user_ids", "=", user.id),
-        ]
-        if profile:
-            # TODO: this probably should be the default only one way
-            # What to do w/ profiles linked to specific user?
-            # This data model is a bit messy :/
-            op_group_domain = [
-                ("operation_group_ids", "in", profile.operation_group_ids.ids),
+        # FIXME exclude menus if they have no picking types
+        # in warehouse matching the profile
```
Comment (guewen) [inline shopfloor/services/menu.py]:
```suggestion
```

File: shopfloor/services/service.py (hunk 5 above)
Comment (guewen) [reply shopfloor/services/service.py:31]: yes, would be worth it I think, but for now I keep it as TODO

File: shopfloor/__manifest__.py (hunk 2 above)
Comment (guewen) [reply shopfloor/__manifest__.py:28]: I'll do it

**Comment by guewen** [comment]
@simahawk rebased

**Pull request merged**